012 - Hot reload feature proposal (#83)
kroxylicious/design: proposals/012-hot-reload-feature.md

This means all virtual clusters are torn down and rebuilt, dropping every client connection even if only one cluster was modified.

In Kubernetes deployments, a configuration change means a pod restart.
In standalone deployments, it means stopping and restarting the process.
Both cause a service interruption that is disproportionate to the scope of the change.

## Motivation

Administrators need to be able to modify proxy configuration in place.
Common scenarios include:

- **Adding or removing virtual clusters** as tenants are onboarded or offboarded.
- **Updating filter configuration** (e.g. updating a KMS endpoint, changing a key selection pattern, modifying ACL rules).
- **Rotating TLS certificates or credentials** that filters reference.

The proxy should apply these changes with minimal disruption: only the virtual clusters affected by the change should experience downtime.
Unaffected clusters should continue serving traffic without interruption.

## Proposal

### Core API: `applyConfiguration()`

The central operation is:

```
applyConfiguration(Configuration)
```

The caller provides a complete `Configuration` object.

> The proxy compares it against the currently running configuration, determines what changed, and applies the changes.

Reviewer: some of these transitions sound interesting from an observability POV. Should we define some metrics in this PR to expose things like how many reloads there have been?

Reviewer 2: Agreed — metrics are part of our public interface and should be defined in this proposal, not deferred as future work. The metrics listed in the current section are the right ones; they should be promoted to a proper part of the proposal.

A configuration apply triggered by an HTTP endpoint should behave identically to one triggered by a file watcher or an operator callback.
These decisions belong in the proxy's static configuration, keeping the `applyConfiguration()` signature simple.

### Configuration change detection

When `applyConfiguration()` is called, the proxy compares the new configuration against the running state to determine which virtual clusters need to be restarted.
A cluster requires a restart if any of the following changed:

- **The virtual cluster model itself** — bootstrap address, TLS settings, gateway configuration, or any other property that contributes to `VirtualClusterModel.equals()`.
- **A filter definition used by the cluster** — if the type or configuration of a `NamedFilterDefinition` referenced by the cluster changed (compared via `equals()`).
- **The default filters list** — if the cluster relies on default filters and the default filters list changed (order matters, since filter chain execution is sequential).

> Clusters where none of these changed are left untouched — they continue serving traffic throughout the apply operation.

Reviewer 3: We should call out explicitly that this means that in-place changes to external files, including keystores, don't be detected. 

I think such a limitation is OK in a first step. It's likely that eventually we will want to support this. It would be worth thinking about that the compatibility story would be at that point -- going from ignoring changes in external files to honouring them would not be a backwards compatible change in behaviour.

### Cluster modification: remove + add

A modified virtual cluster is restarted by tearing it down and rebuilding it with the new configuration.
This is a **remove then add** operation:

1. Gracefully drain existing connections (see below).
2. Deregister the cluster's gateways from the endpoint registry (unbind ports).
3. Register the cluster's new gateways with the endpoint registry (bind ports with new configuration).

> 4. Accept new connections.

Reviewer: Do we need to define this process here now that we have the VCluster lifecycle defined?

Reviewer 2: Agreed — the restructured proposal now defers to [Proposal 016](https://github.com/kroxylicious/design/blob/main/proposals/016-virtual-cluster-lifecycle.md#graceful-shutdown) for drain semantics rather than redefining them.

> 4. Accept new connections.

Reviewer 2: The restructured proposal's orchestration pipeline makes this explicit — validation via the Features framework happens before any VC operations begin. See the pipeline diagram in "Orchestration pipeline."

This means a modified cluster experiences a brief period of unavailability while its ports are unbound and rebound.

> Clients connected to the cluster will be disconnected during the drain phase.

Reviewer 2: This is now covered in [Proposal 016](https://github.com/kroxylicious/design/blob/main/proposals/016-virtual-cluster-lifecycle.md#graceful-shutdown) — default drain timeout is 10s, configurable per virtual cluster. The proposal also acknowledges that long-polling consumers will routinely exceed any reasonable timeout, and that draining is best-effort within the Netty shutdown window.

> This is a deliberate design choice.

Reviewer 2: Agreed, good catch — Uzziee should clarify this.

More surgical approaches — such as swapping the filter chain on existing connections without dropping them, or performing a rolling handoff — would reduce disruption, but they add significant complexity (connection migration, state transfer between filter chain instances, partial rollback of in-flight connections).
The remove+add approach is the right starting point: it is straightforward, predictable, and consistent with how the proxy handles startup failures today.

> The remove+add approach also creates a thundering herd when all disconnected clients reconnect simultaneously after the cluster comes back up; mitigation strategies (e.g. staggered connection acceptance) are future work.

Reviewer 2: Good suggestion.

More surgical alternatives are worth exploring as future work once the foundation is solid.

> Changes are processed in the order: **remove → modify → add**.

Reviewer 3: ```suggestion
When a reconfiguration takes places VMs are classified into one of three cases and are then processed in the order: **remove → modify → add**.
```

Reviewer 2: Good suggestion.

> Removing clusters first frees up ports and resources that new or modified clusters may need.

Reviewer 2: Good suggestion.

### Graceful connection draining

Before tearing down a modified or removed cluster, the proxy drains its connections gracefully rather than dropping them abruptly.
The drain process has three phases:

1. **Reject new connections.** The cluster is marked as draining. Any new client connection attempt to the cluster is immediately refused. Unaffected clusters continue accepting connections normally.

2. **Apply backpressure to existing connections.** On each downstream (client → proxy) channel, reading is disabled (`autoRead = false`) so no new requests are accepted from the client. Upstream (proxy → Kafka) channels continue reading so that responses to already-forwarded requests can flow back to clients.

3. **Wait for in-flight requests to complete, then close.** Each connection is monitored for in-flight Kafka requests. Once the pending request count for a connection reaches zero, the connection is closed. If the count does not reach zero within the drain timeout, the connection is force-closed. The drain timeout should be configurable — long-running consumer rebalances or slow produces with `acks=all` can legitimately exceed a short default.

> This approach ensures that in-progress Kafka operations complete where possible, while bounding the time the proxy waits before proceeding with the restart.

Reviewer 2: Same answer as the drain timeout thread — [Proposal 016](https://github.com/kroxylicious/design/blob/main/proposals/016-virtual-cluster-lifecycle.md#graceful-shutdown) sets the default at 10s, configurable per-VC.

### Failure behaviour and rollback

> The initial default is **all-or-nothing rollback**: if any cluster operation fails during apply (e.g. a port conflict when rebinding, a TLS error, a plugin initialisation failure), all previously successful operations in that apply are rolled back in reverse order.

Reviewer: The VirtualCluster Lifecycle policy talks about a future `onVirtualClusterFailed`, should we update this proposal to define this declarative policy rather than a hardcoded rollback policy

Reviewer 2: See the global comment — the current `onFailure: ROLLBACK | TERMINATE` conflates two independent dimensions. `onVirtualClusterFailed` (per-VC recovery, what 016 anticipated) and `onConfigurationFailure` (orchestration-level rollback) are separate concerns at different levels of abstraction. The global comment has the full decomposition and worked examples.

> The initial default is **all-or-nothing rollback**: if any cluster operation fails during apply (e.g. a port conflict when rebinding, a TLS error, a plugin initialisation failure), all previously successful operations in that apply are rolled back in reverse order.

Reviewer 2: The restructured proposal addresses this — the orchestration pipeline creates the new `FilterChainFactory` (including plugin factory initialization) up-front before any VC changes are processed. Failures at that stage prevent any VC operations from starting.

Added clusters are removed, modified clusters are reverted to their original configuration, and removed clusters are re-added.

This is consistent with startup behaviour, where a failure in any virtual cluster fails the entire proxy.

> It produces a predictable outcome: after a failed apply, the proxy is in its previous known-good state, and the administrator can investigate and retry.

Reviewer: Does this section need to be updated in light of the Virtual Cluster Lifecycle? Once we have per-VC lifecycles this won't be the startup behaviour will it?

Reviewer 2: See the global comment — this proposal is where per-VC independence becomes meaningful. 016 defined the model; this proposal defines the recovery and failure policies that make it operational.

The change detection described above can identify when a filter's YAML configuration changes (via `equals()` on the configuration model), but it cannot detect when external resources that a plugin reads during initialisation have changed.
For example, a password file, TLS keystore, or ACL rules file may have changed on disk even though the plugin's configuration (which only references the file path) is identical.

These reads typically happen deep in nested plugin call stacks (e.g. `RecordEncryption` → `KmsService` → `CredentialProvider` → `FilePassword`), so the runtime has no visibility into what was read or whether it has changed.

Without addressing this gap, an apply operation would miss these changes entirely — the plugin configuration hasn't changed, so no restart is triggered, even though the plugin's actual behaviour would differ if it were reinitialised.

An approach is being explored where plugins read external resources through the runtime (rather than doing direct file I/O), allowing the runtime to track what was read and hash the content for subsequent change detection.
This makes dependency tracking automatic rather than relying on plugin authors to opt in.

> The detailed design for this mechanism will be covered in a separate proposal to keep this one focused on the core apply operation.

Reviewer 2: The restructured proposal covers this in the "Plugin resource tracking (known gap)" section and defers the detailed design to a separate proposal. Happy to go deeper on the approach if useful — the thinking is outlined in the PR comment history.

## Open questions

> - **Configuration granularity**: The initial design uses state-of-the-world snapshots. Is there a use case that requires delta-based operations or more targeted snapshots in the near term, or is this purely future work?

Reviewer 2: Agreed — state of the world is the right interface. Implementation can optimise internally.

> - **Failure behaviour options beyond all-or-nothing**: What specific deployment models need partial-success or terminate-on-failure semantics, and what configuration surface do they need?

Reviewer 2: See the global comment — decomposing `ROLLBACK | TERMINATE` into two independent dimensions (rollback + serve) actually reveals a use case for non-all-or-nothing: security/compliance enforcement where the old config is unacceptable (e.g. enforcing new TLS cipher suites). In that case you'd rather have a partial deployment than revert to insecure config. The decomposition surfaces this naturally without us having to design for it specially.

> - **Drain timeout default and configurability**: What is a reasonable default drain timeout? How should it be configured — globally, per-cluster, or both?

Reviewer 2: [Proposal 016](https://github.com/kroxylicious/design/blob/main/proposals/016-virtual-cluster-lifecycle.md#graceful-shutdown) landed on best-effort draining — favour a short drain time, Kafka clients handle connection loss, and idempotent producers retain their PID/epoch across reconnections. Agreed that transparent cluster switching is a separate, harder problem and shouldn't block this.

**Trigger mechanisms are explicitly out of scope for this proposal.** The `applyConfiguration()` operation is the internal interface that any trigger plugs into. How the new configuration arrives — whether via an HTTP endpoint, a file watcher detecting a changed ConfigMap, or a Kubernetes operator callback — is a separate concern. Deferring this keeps the proposal focused and avoids blocking on unresolved questions about trigger design (see [Trigger mechanisms](#trigger-mechanisms-future-work) below).

**Failure behaviour is deployment-level static configuration.** Whether the proxy rolls back or terminates on failure is controlled by `ReloadOptions`, defined once in the proxy's static YAML configuration file. This ensures consistent, operator-controlled behaviour regardless of trigger mechanism. A configuration apply triggered by an HTTP endpoint should behave identically to one triggered by a file watcher or an operator callback. These decisions belong in the proxy's static configuration, keeping the `applyConfiguration()` signature simple.

```yaml
reloadOptions:
  onFailure: ROLLBACK          # ROLLBACK | TERMINATE

>   persistConfigToDisk: true    # true | false

Reviewer 4: is it applicable for failure and success scenario?

Author: This is stale, check the latest matrix with all the scenarios

| | `false` | Don't persist. The reload is in-memory only.                                                           |

### Configuration change detection

When `applyConfiguration()` is called, the proxy compares the new configuration against the running state to determine which virtual clusters need to be restarted. Change detection is implemented as a pipeline of `ChangeDetector` implementations, each responsible for one category of change:

- **`VirtualClusterChangeDetector`** — identifies clusters that were added, removed, or modified by comparing `VirtualClusterModel` instances via `equals()`. A cluster requires a restart if any property that contributes to `VirtualClusterModel.equals()` changed (bootstrap address, TLS settings, gateway configuration, etc.).
- **`FilterChangeDetector`** — identifies clusters affected by filter configuration changes. A cluster requires a restart if a `NamedFilterDefinition` it references changed (type or configuration, compared via `equals()`), or if the `defaultFilters` list changed (order matters, since filter chain execution is sequential) and the cluster relies on default filters.

> Detectors return a `ChangeResult(clustersToRemove, clustersToAdd, clustersToModify)`. Results from all detectors are aggregated via `LinkedHashSet` to maintain order while deduplicating cluster names that appear in multiple detector results.

Reviewer 4: nit:
```suggestion
Detectors return a `ChangeResult(deleted, added, updated)`. Results from all detectors are aggregated via `LinkedHashSet` to maintain order while deduplicating cluster names that appear in multiple detector results.
```

> Detectors return a `ChangeResult(clustersToRemove, clustersToAdd, clustersToModify)`. Results from all detectors are aggregated via `LinkedHashSet` to maintain order while deduplicating cluster names that appear in multiple detector results.

Reviewer 4: Results from all detectors are aggregated via `LinkedHashSet`  - this is implementation detail, we can remove this and keep that changes will be aggregated at virtual-cluster or vc-gateway level.

Clusters where none of these changed are left untouched — they continue serving traffic throughout the apply operation.

### Cluster modification via lifecycle transitions

A modified virtual cluster is restarted by driving it through the lifecycle states defined in Proposal 016. Proposal 016 defines the per-VC state machine (`VirtualClusterLifecycleState`) and the `VirtualClusterLifecycleManager` that enforces valid transitions. This proposal adds the reload operations that drive those transitions.

> The three reload operations map to lifecycle transitions as follows:

Reviewer 4: nit: 
```suggestion
The three change operations map to lifecycle transitions as follows:
```

> **Restart (modify):** `SERVING → DRAINING → [drain connections] → [deregister gateways] → INITIALIZING → [register gateways] → SERVING`

Reviewer 4: nit:
```suggestion
**Modify (Restart VC/VC-GW):** `SERVING → DRAINING → [drain connections] → [deregister gateways] → INITIALIZING → [register gateways] → SERVING`
```
rationale - avoids confusion

A modified cluster is torn down and rebuilt with the new configuration. During restart, the lifecycle state cycles through Draining and back to Initializing without ever reaching the terminal Stopped state. This means the `onVirtualClusterStopped` callback (defined in Proposal 016) does not fire during restart — reload is an internal VCM operation that stays within the lifecycle state machine.

**Remove:** `SERVING → DRAINING → [drain connections] → [deregister gateways] → STOPPED`

> A removed cluster is permanently torn down. It reaches the terminal Stopped state, and the `onVirtualClusterStopped` callback fires. If the proxy's `onVirtualClusterStopped` failure policy is `serve: none` (the default from Proposal 016), the proxy shuts down. If the policy is `serve: successful`, the proxy continues with the remaining healthy virtual clusters.

Reviewer 4: explicit callout - `serve` itself might get changed, so the check should be performed using the newConfiguration.

Author: We are only detecting changes to either virtual cluster or filter configs, the `serve` config does not belong to either of these. Even if the operator changes the `serve` config, it wont be detected in this version.

- **`kroxylicious_reload_clusters_affected_total`** — counter of per-VC operations during reload, labelled by operation (`add`, `remove`, `modify`) and outcome (`success`, `failure`, `rolledback`). Provides granularity beyond the aggregate reload result.
- **`kroxylicious_drain_duration_seconds`** — histogram of per-VC connection drain duration. Helps tune the `drainTimeout` configuration and detect VCs with long-lived connections.
- **`kroxylicious_drain_connections_force_closed_total`** — counter of connections force-closed after drain timeout. A high rate indicates the drain timeout is too aggressive for the workload.

The per-VC lifecycle state metrics (`kroxylicious_virtual_cluster_state`, `kroxylicious_virtual_cluster_state_duration_seconds`, `kroxylicious_virtual_cluster_transitions_total`) defined in Proposal 016 complement these reload-specific metrics and should be implemented alongside them.

A management endpoint exposing the last reload result (timestamp, outcome, affected clusters, duration) would also be valuable for on-demand inspection by operators. The design of this endpoint is deferred along with the metrics implementation.

## Open questions

> - **Rollback and `onVirtualClusterStopped` interaction**: If a remove operation succeeds (VC reaches Stopped, callback fires) but a subsequent add operation fails and triggers rollback (re-adding the removed VC), the callback will have already fired for the Stopped transition. Should the rollback suppress or compensate for the callback? With `serve: none`, the callback may trigger proxy shutdown before rollback has a chance to restore the VC.

Reviewer: This seems like a VCLC problem to me, that the proxy-level `serve` policy was designed to address failures rather than an intentional removal of a cluster. If I have VC-A and VC-B and remove VC-A, do I want the VC to shutdown on successful removal? I don't think we do.

Maybe the policy should be `onVirtualClusterTerminallyFailed` and we remodel so there are either two terminal states or make the stopped state aware of why it stopped? WDYT @SamBarker

Author: Yup, this is what I was thinking

Make `Stopped` reason-aware by adding a `StopReason` to the Stopped record:

```
enum StopReason { SHUTDOWN, REMOVED, FAILED }
                                                                                                                                                                                                                 
record Stopped(StopReason reason, @Nullable Throwable priorFailureCause)                                                                                                                                       
      implements VirtualClusterLifecycleState {}     
```

The `serve` policy only triggers for `StopReason.FAILED`. The callback stays the same but the consumer inspects the reason

```
// In VCM:                                                                                                                                                                                                     
manager.stop();  // Failed → Stopped(FAILED, cause)                                                                                                     
manager.drainComplete();  // Draining → Stopped(SHUTDOWN, null)                                                                                                                                                
                                                                                                                                                                                                                 
// In hot-reload:                                                                                                                                                                                              
manager.drainComplete();  // Draining → Stopped(REMOVED, null)
```

Reviewer 2: Good catch — I genuinely didn't consider intentional removal as a path to `stopped` when writing 016. You're right that the `serve` policy conflates "failed and unrecoverable" with "intentionally removed," and those need different treatment.

I've been thinking about this with @Uzziee's `StopReason` suggestion as a starting point, and I think the fix is slightly different. The distinction isn't about the *state* — a stopped cluster is permanently done regardless of why. The distinction is about which *transition* got it there.

In graph theory terms: move the policy hook from the node to the edge. The `serve` policy should fire on the `failed → stopped` edge, not on arrival at `stopped`. The `draining → stopped` edge (removal, shutdown) doesn't trigger it.

- Rename `onVirtualClusterStopped` → `onVirtualClusterTerminalFailure`
- The state model stays unchanged — one terminal state, same transitions
- During shutdown, the proxy-level owner suppresses the callback (it's already shutting down) — that's a caller concern, not a state machine concern

This also resolves the open question about rollback interaction: an intentional removal reaches `stopped` via `draining → stopped`, the failure policy doesn't fire, so there's no race between the callback triggering proxy shutdown and the rollback trying to restore the VC.

This refinement belongs in the hot reload proposal since that's where the need was discovered — 016 was a point-in-time snapshot.

# Changing Active Proxy Configuration

**Builds on:** [Proposal 016 — Virtual Cluster Lifecycle](https://github.com/kroxylicious/design/blob/main/proposals/016-virtual-cluster-lifecycle.md)

> This proposal introduces a mechanism for applying configuration changes to a running Kroxylicious proxy without a full restart. It defines a core `applyConfiguration(Configuration)` operation that accepts a complete configuration, detects what changed, and converges the running state to match — restarting only the affected virtual clusters while leaving unaffected clusters available.

Reviewer 3: It's a good idea to be explicit right from the start about which class you're proposing to add this method to:

```suggestion
This proposal introduces a mechanism for applying configuration changes to a running Kroxylicious proxy without a full restart. It defines a core `KafkaProxy.applyConfiguration(Configuration)` operation that accepts a complete configuration, detects what changed, and converges the running state to match — restarting only the affected virtual clusters while leaving unaffected clusters available.
```

In Kubernetes deployments, a configuration change means a pod restart. In standalone deployments, it means stopping and restarting the process. Both cause a service interruption that is disproportionate to the scope of the change.

## Motivation

Administrators need to be able to modify proxy configuration in place. Common scenarios include:

- **Adding or removing virtual clusters** as tenants are onboarded or offboarded.
- **Updating filter configuration** (e.g. updating a KMS endpoint, changing a key selection pattern, modifying ACL rules).

The proxy should apply these changes with minimal disruption: only the virtual clusters affected by the change should experience downtime. Unaffected clusters should continue serving traffic without interruption.

This proposal also delivers the startup behaviour change that Proposal 016 made possible. With `serve: successful` and reload, an operator can let healthy VCs serve traffic while fixing a broken VC's config and re-applying — making per-VC independence useful rather than theoretical.

## Proposal

### Core API: `applyConfiguration()`

The central operation is:

```java

> public CompletableFuture<ReloadResult> applyConfiguration(Configuration newConfig)

Reviewer 3: Again, make it clear exactly what change you're proposing:

```suggestion
class KafkaProxy {
    // ... add the following method
    
    public CompletableFuture<ReloadResult> applyConfiguration(Configuration newConfig)
}
```

It would be a good idea to provide the contract for this method as Javadoc, in particular:
* Can the method throw, or are all exceptional conditions exposed via exceptional completion of the returned future?
* Are there any exceptional conditions which are expected and worth documenting?
* You should also explain `ReloadResult`. `ReloadResult` is currently never mentioned again in this proposal. Something called `ChangeResult` _is_ mentioned is that the same thing?

This method accepts the `Configuration`, we can assume it has already been parsed and some validation has happened. But you should be very clear whether you're expecting all validation to have already been done, or whether `applyConfiguration` might perform any checks of its own. I _think_ we might be inching towards saying something like:

> Static validation (i.e. based on the configuration files alone) is expected to have already been done before this method is called. Validation which depends on runtime state (like port conflicts) will be done during `applyConfiguration()` and reported through the `ReloadResult`

Getting agreement about this contract, and ensuring the proposal communicates it clearly, is a key part of the proposal process.

Author: Have added a javadoc which mentions when will the method throw and other exceptional conditions.

`ReloadResult` - I have documented this now. The invoker of this API will use this result and make further decisions.

`ChangeResult` - this is just an intermediary object to hold the list of virtual clusters which are to be either added, removed or restarted based on the new config we are trying to apply. This is internal to proxy.

```

The caller provides a complete `Configuration` object. The proxy compares it against the currently running configuration, determines what changed, and applies the changes. The method returns a `CompletableFuture<ReloadResult>` that completes with a structured result on success or exceptionally on failure.

> In this approach: the caller provides the desired end state, and the proxy is responsible for computing and executing the diff. This is the right starting point — it is simple to reason about and avoids the complexity of delta-based or partial-update APIs. More granular approaches (deltas, targeted snapshots) are worth exploring later, but the initial API should leave room for them without committing to them now.

Reviewer 3: ```suggestion
In this approach: the caller provides the desired end state, and the proxy is responsible for computing and executing the diff. This is the right starting point — it is simple to reason about and avoids the complexity of delta-based or partial-update APIs. More granular approaches (deltas, targeted snapshots) may be worth exploring later, but the initial API should leave room for them without committing to them now.
```

  }

  
}
```

3. **ConfigurationChangeRollbackTracker** - This class maintains a record of all cluster operations (removals, modifications, additions) so they can be reversed if the overall configuration change fails.
```
public class ConfigurationChangeRollbackTracker {
  
    /**
     * Tracks a cluster removal operation.
     */
    public void trackRemoval(String clusterName, VirtualClusterModel removedModel) {}

    /**
     * Tracks a cluster modification operation.
     */
    public void trackModification(String clusterName, VirtualClusterModel originalModel, VirtualClusterModel newModel) {}

    /**
     * Tracks a cluster addition operation.
     */
    public void trackAddition(String clusterName, VirtualClusterModel addedModel) {}
}
```

4. **ChangeDetector (Interface)** - Strategy pattern interface for different types of change detection.
    - Currently we have only have one implementation - `VirtualClusterChangeDetector` which will detect changes in virtual cluster models, in future we can add a detector for filter changes aka `FilterChangeDetector`
    - Provides a consistent API for comparing old vs new configurations via detectChanges()

>     - Returns structured ChangeResult objects with specific operations needed

Reviewer 5: Does this truly have to be pluggable? Is there any actual difference in terms of finding changes in different parts of the config tree? Intuitively, I'd expect a single implementation which identifies changes to the config tree, and these changes then may be processed by a pluggable set of config change handlers (and there might be changes which cannot be processed yet as there's no handler acting on them).

    
    private static final Logger LOGGER = LoggerFactory.getLogger(VirtualClusterChangeDetector.class);
    
    @Override
    public String getName() {
        return "VirtualClusterChangeDetector";
    }
    
    @Override
    public ChangeResult detectChanges(ConfigurationChangeContext context) {
        // Check for modified clusters using equals() comparison
        List<String> modifiedClusters = findModifiedClusters(context);
        
        // Check for new clusters (exist in new but not old)
        List<String> newClusters = findNewClusters(context);
        
        // Check for removed clusters (exist in old but not new)
        List<String> removedClusters = findRemovedClusters(context);
        
        return new ChangeResult(removedClusters, newClusters, modifiedClusters);
    }
```

5. **ConfigurationChangeContext (Record)** - Immutable data container providing context for change detection.
    - Holds old and new Configuration objects for comparison
```
public record ConfigurationChangeContext(
    Configuration oldConfig,
    Configuration newConfig,
    List<VirtualClusterModel> oldModels,
    List<VirtualClusterModel> newModels

> ) {}

Reviewer 5: How will that accommodate changes to config parts other than clusters?

```

6. **ChangeResult (Record)** - Contains lists of cluster names for each operation type (remove/add/modify)
```
public record ChangeResult(
    List<String> clustersToRemove,
    List<String> clustersToAdd, 
    List<String> clustersToModify
) {}
```

## Flow diagram

> <img width="2354" height="1394" alt="Image" src="https://github.com/user-attachments/assets/19a7edba-8881-4fa7-98c8-defcdbb132da" />

Reviewer 5: This doesn't render.

```suggestion
[Flow Diagram](https://github.com/user-attachments/assets/19a7edba-8881-4fa7-98c8-defcdbb132da)
```
